### PR TITLE
Start dynamic heap count work

### PR DESCRIPTION
**Pull request by PeterSolMS**
This is work to enable changing the heap count for server GC dynamically:
- move regions between heaps
- make handle table scanning accommodate varying numbers of GC threads
- changes to mark list and mark list piece sizing
- rethreading free lists for gen 2 (other generations to follow)
- update thread numbers in joins to make joins work correctly
- park idle GC threads on a new event `gc_idle_thread_event`
- changes to allow `try_allocate_more_space` to switch to a different heap
- move finalization data between heaps

Note that with the current code we are always running in a stress mode where we vary the heap count randomly every 100 GCs.

### CODE REVIEW

**Comment by Maoni0** [comment]
>     gc_join_max = 40

this should be deleted since you are not using it.

---
Refers to: src/coreclr/gc/gc.cpp:704 in fa4cf11. [](commit_id = fa4cf11a6c39eb6119db8a7033e3b627330003cc, deletion_comment = False)

File: src/coreclr/debug/inc/amd64/primitives.h [hunk 1]
```diff
@@ -70,7 +70,9 @@ constexpr CorDebugRegister g_JITToCorDbgReg[] =
     REGISTER_AMD64_R12,
     REGISTER_AMD64_R13,
     REGISTER_AMD64_R14,
-    REGISTER_AMD64_R15
+    REGISTER_AMD64_R15,
+    REGISTER_AMD64_RSP,
+    REGISTER_AMD64_RSP,
```
Comment (Maoni0) [inline src/coreclr/debug/inc/amd64/primitives.h:75]: is this related to this PR?

File: src/coreclr/gc/gcinterface.h [hunk 2]
```diff
@@ -1032,6 +1032,7 @@ struct ScanContext
 {
     Thread* thread_under_crawl;
     int thread_number;
+    int thread_count;
```
Comment (Maoni0) [inline src/coreclr/gc/gcinterface.h:1035]: this would break the backward compatibility with the VM side (ie, you can't run clrgc.dll with a previous version of coreclr.dll anymore). can you change it in a way that does not introduce this breaking change?

File: src/coreclr/gc/gcpriv.h [hunk 3]
```diff
@@ -142,6 +142,14 @@ inline void FATAL_GC_ERROR()
 #define USE_REGIONS
 #endif //HOST_64BIT && BUILD_AS_STANDALONE
 
+#define SPINLOCK_HISTORY
+#define RECORD_LOH_STATE
```
Comment (Maoni0) [inline src/coreclr/gc/gcpriv.h:146]:
```suggestion
// #define SPINLOCK_HISTORY
// #define RECORD_LOH_STATE
```

Comment (Maoni0) [reply src/coreclr/gc/gcpriv.h:146]: you wouldnt want to define these unconditionally - at least not in release builds

File: src/coreclr/gc/gcpriv.h [hunk 4]
```diff
@@ -347,7 +355,8 @@ const int policy_expand  = 2;
 #ifdef SIMPLE_DPRINTF
 
 void GCLog (const char *fmt, ... );
-#define dprintf(l,x) {if ((l == 1) || (l == GTC_LOG)) {GCLog x;}}
+//#define dprintf(l,x) {if ((l == 1) || (l == GTC_LOG)) {GCLog x;}}
+#define dprintf(l,x) {if ((l == 1) || (l == 5555)) {GCLog x;}}
```
Comment (Maoni0) [inline src/coreclr/gc/gcpriv.h:359]:
```suggestion
#define dprintf(l,x) {if ((l == 1) || (l == GTC_LOG)) {GCLog x;}}
```

File: src/coreclr/gc/gcpriv.h [hunk 5]
```diff
@@ -571,9 +580,17 @@ enum allocation_state
     a_state_trigger_ephemeral_gc,
     a_state_trigger_2nd_ephemeral_gc,
     a_state_check_retry_seg,
+    // a temp state I added to distinguish 
+    a_state_wait_in_tams,
```
Comment (Maoni0) [inline src/coreclr/gc/gcpriv.h:584]: did you want to get rid of this?

File: src/coreclr/gc/gcpriv.h [hunk 6]
```diff
@@ -3927,13 +4013,19 @@ class gc_heap
...
 
     // Also updated on the heap#0 GC thread because that's where we are actually doing the decommit.
     PER_HEAP_ISOLATED_FIELD_SINGLE_GC BOOL gradual_decommit_in_progress_p;
 #ifdef MH_SC_MARK
     PER_HEAP_ISOLATED_FIELD_SINGLE_GC int* g_mark_stack_busy;
 #endif //MH_SC_MARK
 
+#ifdef USE_REGIONS
+    PER_HEAP_ISOLATED_FIELD_SINGLE_GC size_t total_num_fl_items_stage1;
+    PER_HEAP_ISOLATED_FIELD_SINGLE_GC size_t total_num_fl_items_moved_stage1;
```
Comment (Maoni0) [inline src/coreclr/gc/gcpriv.h:4026]: I wrote these for verification purposes. you could get rid of them.

File: src/coreclr/gc/gcpriv.h [hunk 7]
```diff
@@ -4391,6 +4485,10 @@ class gc_heap
     PER_HEAP_ISOLATED_FIELD_DIAG_ONLY BOOL heap_analyze_enabled;
 #endif //HEAP_ANALYZE
 
+#ifdef MULTIPLE_HEAPS
+    PER_HEAP_FIELD bool uoh_msl_before_gc_p;
```
Comment (Maoni0) [inline src/coreclr/gc/gcpriv.h:4489]: this should be under the `#ifdef STRESS_DYNAMIC_HEAP_COUNT` check shoiuldn't it?

File: src/coreclr/gc/gc.cpp [hunk 8]
```diff
@@ -1376,6 +1385,85 @@ static void safe_switch_to_thread()
...
+    }
+
+static const int32_t lock_free = -1;
+static const int32_t lock_taken = 0;
+static const int32_t lock_decommissioned = 1;
+
+
+// If our heap got decommissioned, we need to try an existing heap.
+//inline
+#pragma optimize("", off)
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:1401]:
```suggestion
```

File: src/coreclr/gc/gc.cpp [hunk 9]
```diff
@@ -1376,6 +1385,85 @@ static void safe_switch_to_thread()
...
+    if (msl->lock == lock_decommissioned)
+    {
+        dprintf (5555, ("heap#%d got decommissioned! need to retry", heap_number));
+    }
+    return (msl->lock == lock_decommissioned);
+#else //MULTIPLE_HEAPS
+    return false;
+#endif //MULTIPLE_HEAPS
+}
+#pragma optimize("", on)
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:1414]:
```suggestion
```

Comment (Maoni0) [reply src/coreclr/gc/gc.cpp:1414]: I turned the optimizatio off for debugging purpose. should not be checked in.

File: src/coreclr/gc/gc.cpp [hunk 10]
```diff
@@ -1385,11 +1473,12 @@ static void enter_spin_lock_noinstru (RAW_KEYWORD(volatile) int32_t* lock)
 {
 retry:
 
-    if (Interlocked::CompareExchange(lock, 0, -1) >= 0)
+    if (Interlocked::CompareExchange(lock, lock_taken, lock_free) != lock_free)
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:1476]: this method doesn't actually need to change, correct? it will work as is. but I do think it's better to use the actualy enum instead of an integer. perhaps add a comment to say that this will never be used by the more space lock?

File: src/coreclr/gc/gc.cpp [hunk 11]
```diff
@@ -6885,11 +7032,26 @@ void gc_heap::gc_thread_function ()
...
+                if ((lowest_heap_with_msl_uoh == -1) && (hp->uoh_msl_before_gc_p))
+                {
+                    lowest_heap_with_msl_uoh = i;
+                }
+
+                if (hp->uoh_msl_before_gc_p)
+                {
+                    dprintf (5555, ("h%d uoh msl was taken before GC", i));
+                    hp->uoh_msl_before_gc_p = false;
+                }
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:7054]: shouldn't this code be under `#ifdef STRESS_DYNAMIC_HEAP_COUNT`?

File: src/coreclr/gc/gc.cpp [hunk 12]
```diff
@@ -1581,19 +1676,24 @@ static void enter_spin_lock (GCSpinLock* spin_lock)
...
 {
-    spin_lock->lock = -1;
+    spin_lock->lock = lock_free;
 }
 
 #define ASSERT_HOLDING_SPIN_LOCK(pSpinLock)
 
 #endif //_DEBUG
 
+inline static BOOL spin_lock_taken_p (GCSpinLock* pSpinLock)
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:1692]: do we need to 2 `spin_lock_taken_p` functions?

File: src/coreclr/gc/gc.cpp [hunk 13]
```diff
@@ -2215,6 +2319,10 @@ size_t*     gc_heap::g_bpromoted;
 
 BOOL        gc_heap::gradual_decommit_in_progress_p = FALSE;
 size_t      gc_heap::max_decommit_step_size = 0;
+#ifdef USE_REGIONS
+size_t      gc_heap::total_num_fl_items_moved_stage1 = 0;
+size_t      gc_heap::total_num_fl_items_stage1 = 0;
+#endif //USE_REGIONS
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:2325]:
```suggestion
```

File: src/coreclr/gc/gc.cpp [hunk 14]
```diff
@@ -2225,6 +2333,8 @@ size_t      gc_heap::g_promoted;
 size_t      gc_heap::g_bpromoted;
 #endif //BACKGROUND_GC
 
+const int n_heaps = 1;
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:2336]: can you please add an explanation why this is called the same name as the one in `gc_heap`?

File: src/coreclr/gc/gc.cpp [hunk 15]
```diff
@@ -6217,7 +6339,12 @@ class heap_select
         if (GCToOSInterface::CanGetCurrentProcessorNumber())
         {
             uint32_t proc_no = GCToOSInterface::GetCurrentProcessorNumber();
-            return proc_no_to_heap_no[proc_no];
+            int best_heap = proc_no_to_heap_no[proc_no];
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:6342]: I wouldn't call this `best_heap` which makes it sound like this is just a perf thing. but it wouldn't work if we didn't do `%`. probably call it something like `adjusted_heap`.

File: src/coreclr/gc/gc.cpp [hunk 16]
```diff
@@ -23562,7 +24282,21 @@ void gc_heap::equalize_promoted_bytes()
...
+            gc_heap* hp = g_heaps[i];
+
+            generation* gen = hp->generation_of (gen_idx);
+
+            // if the number of heaps changes, some heaps may end up with
+            // no regions at all - repair this here
+            if (heap_segment_rw (generation_start_segment (gen)) == nullptr)
+            {
+                size_t size = gen_idx > max_generation ? global_region_allocator.get_large_region_alignment() : 0;
+                heap_segment* start_region = hp->get_free_region (gen_idx, size);
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:24294]: this can fail - at the minimum we should add a TODO here.

File: src/coreclr/gc/gc.cpp [hunk 17]
```diff
@@ -23578,8 +24312,677 @@ void gc_heap::equalize_promoted_bytes()
...
+    spinlock_info_index = 0;
+    current_uoh_alloc_state = (allocation_state)-1;
+#endif //SPINLOCK_HISTORY
+
+#ifdef RECORD_LOH_STATE
+    loh_state_index = 0;
+#endif //RECORD_LOH_STATE
+}
+
+bool gc_heap::change_heap_count (int new_n_heaps)
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:24748]: it would be great if you could add some detailed comments in this method like for each thing we are adjusting, what we are actually doing, like what we do in `thread_final_regions` or `equalize_promoted_bytes`, eg

```
// step 1: adjust finalize queue. if the heap count is larger, [describe how we split the old heaps' finalize queue into more queues]; if it's smaller, [describe how we decide to merge which heap into which heap]
// step 2: for all existing regions, prepare the regions by pretending all their allocated space survives so equalize_promoted_size will distribute regions evenly across new heaps
// ...
```
so we don't just say "adjust data for new heaps" but rather describe how we are adjusting - for example, if we go from 10 to 6, are we adjusting by taking the last 4 heaps and move all their regions to the 1st 4 heaps? or are we distributing evenly to all 6 heaps?

File: src/coreclr/gc/gc.cpp [hunk 18]
```diff
@@ -23578,8 +24312,677 @@ void gc_heap::equalize_promoted_bytes()
...
+                // failing to merge finalization data from one of the heaps about to go idle
+                // means we cannot in fact reduce the number of heaps.
+                dprintf (3, ("failed to merge finalization from heap %d into heap %d", i, to_heap_number));
+                return false;
+            }
+
+            to_heap_number = (to_heap_number + 1) % new_n_heaps;
+        }
+    }
+
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:24777]: from this point on, is it true we cannot fail? (it's actually not true since we could fail in equalize_promoted_bytes. is it ok do all the steps that can fail before doing steps that can't?

File: src/coreclr/gc/gc.cpp [hunk 19]
```diff
@@ -23578,8 +24312,677 @@ void gc_heap::equalize_promoted_bytes()
...
+#ifdef BACKGROUND_GC
+    bgc_t_join.update_n_threads(new_n_heaps);
+#endif //BACKGROUND_GC
+
+    // even out the regions over the current number of heaps
+    equalize_promoted_bytes (max_generation);
+
+    total_num_fl_items_moved_stage1 = ~0;
+    total_num_fl_items_stage1 = ~0;
+
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:24922]:
```suggestion
```

File: src/coreclr/gc/gc.cpp [hunk 20]
```diff
@@ -23578,8 +24312,677 @@ void gc_heap::equalize_promoted_bytes()
...
+    {
+        gc_heap* hp = g_heaps[i];
+
+        hp->decommission_heap();
+    }
+
+    // make sure no allocation contexts point to idle heaps
+    fix_allocation_contexts_heaps();
+
+    GCConfig::s_HeapVerifyLevel = 1;
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:24978]: you could just call verify_heap here instead of turning it on for all GCs later as we discussed (or verify this one and a couple of GCs later

File: src/coreclr/debug/inc/amd64/primitives.h (hunk 1 above)
Comment (PeterSolMS) [reply src/coreclr/debug/inc/amd64/primitives.h:75]: No, this is just a temporary hack to work around a debugging issue I encountered - I'll take it out.

File: src/coreclr/gc/gcpriv.h (hunk 3 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gcpriv.h:146]: Will do.

File: src/coreclr/gc/gcpriv.h (hunk 4 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gcpriv.h:359]: Sure...

File: src/coreclr/gc/gcpriv.h (hunk 5 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gcpriv.h:584]: Yes, this seems low value at this point.

File: src/coreclr/gc/gc.cpp (hunk 10 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:1476]: Right, this doesn't need a change, I just did it for clarity. Will add a comment as suggested.

File: src/coreclr/gc/gc.cpp (hunk 11 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:7054]: Yes, it should be.

File: src/coreclr/gc/gc.cpp (hunk 13 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:2325]: Sure.

File: src/coreclr/gc/gc.cpp (hunk 19 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:24922]: Sure, I'll remove the declarations and uses.

File: src/coreclr/gc/gc.cpp (hunk 14 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:2336]: I added a comment to that effect as well.

File: src/coreclr/gc/gc.cpp (hunk 15 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:6342]: ok.

File: src/coreclr/gc/gc.cpp (hunk 16 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:24294]: I have added a comment.

File: src/coreclr/gc/gc.cpp (hunk 18 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:24777]: I think we absolutely need to avoid the case where we get stuck because we failed in the middle of adjusting the heap count, but cannot go back. One way to avoid that is to do all the steps that can fail first, and make sure we can recover. I think this is feasible, but of course you are correct in pointing out it's curently not the case.

File: src/coreclr/gc/gc.cpp (hunk 20 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:24978]: Hmm, this code is on a single thread, so I'd have to iterate over all heaps. Needs more thought.

File: src/coreclr/gc/gc.cpp (hunk 17 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:24748]: Will do...

File: src/coreclr/gc/gc.cpp (hunk 16 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp]: I rearranged the code so we check before whether we have or can get enough regions for each generation **before** we commit to increasing the heap count.

File: src/coreclr/gc/gc.cpp (hunk 20 above)
Comment (Maoni0) [reply src/coreclr/gc/gc.cpp:24978]: ahh right, one idea is to set a global var to indicate that we did change the HC (heap count) like -

`PER_HEAP_ISOLATED_FIELD_SINGLE_GC bool last_suspension_hc_changed_p; `

and next time we do a GC, we will call `verify_heap` on entry if this is true and that GC will reset it. or you could keep a var to record the GC index when we suspended to change the HC, and call `verify_heap` in the next few GCs if this is not -1.

Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:24978]: Actually, in the dynamic heap count PR building on this one, I execute the `change_heap_count` method on all GC threads, and so it would be easy to just call `verify_heap` at the end.

**Comment by PeterSolMS** [comment]
Closing this pull request as it has been superseded by the dynamic heap count one.
